### ISSUE TITLE
Different behavior between `Union` and `|` in mypy when dealing with optionals

### ISSUE THREAD

**Issue by jevandezande** [python/mypy#10499]
In the following two code examples, the first gives an error, while the latter does not. As I understand `type1 | type2` is supposed to be an alias for `Union[type1, type2]` in python3.10 (hence the `from __future__ import annotations`).

```python
from __future__ import annotations

def f(a: int | str = None):
    pass
```
`error: Incompatible default for argument "a" (default has type "None", argument has type "Union[int, str]")`

```python
from typing import Union

def f(a: Union[int, str] = None):
    pass
```

As of [PEP 484](https://www.python.org/dev/peps/pep-0484/#union-types), non-explicit optional is discouraged. Is this the correct behavior, or a bug?

**Your Environment**

- python = `3.9.1`
- mypy = `0.812`

**Comment by JelleZijlstra** [comment]
This is related to the [--no-implicit-optional](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional) flag. We're already planning to flip the default for this flag (#9091), so maybe it's a good thing that the special casing doesn't apply to the new `|` operator.

**Comment by hauntsaninja** [comment]
It wasn't intentional that implicit optionals don't work with PEP 604 syntax, but I'm not going to add the special casing for it. As you say, PEP 484 discourages this, it's not really supported by other type checkers, and the next release of mypy flips the default for `--no-implicit-optional`.
